From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Eric Kilmer <eric.d.kilmer@gmail.com>
Date: Wed, 24 May 2023 19:03:25 -0400
Subject: [PATCH] 5382: Fix ASAN static initialization order fiasco

See here for more details
https://github.com/google/sanitizers/wiki/AddressSanitizerInitializationOrderFiasco

Use the "Construct On First Use" idiom from
https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Construct_On_First_Use
---
 Ghidra/Features/Decompiler/src/decompile/cpp/test.cc | 4 +---
 Ghidra/Features/Decompiler/src/decompile/cpp/test.hh | 9 +++++++--
 2 files changed, 8 insertions(+), 5 deletions(-)

diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/test.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/test.cc
index 660f069400..603aaa5bb8 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/test.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/test.cc
@@ -18,8 +18,6 @@
 
 namespace ghidra {
 
-vector<UnitTest *> UnitTest::tests;
-
 /// Run all the tests unless a non-empty set of names is passed in.
 /// In which case, only the named tests in the set are run.
 /// \param testNames is the set of names
@@ -30,7 +28,7 @@ int UnitTest::run(set<string> &testNames)
   int total = 0;
   int passed = 0;
 
-  for(auto &t : UnitTest::tests) {
+  for(auto &t : UnitTest::tests()) {
     if (testNames.size() > 0 && testNames.find(t->name) == testNames.end()) {
       continue;
     }
diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/test.hh b/Ghidra/Features/Decompiler/src/decompile/cpp/test.hh
index 27c27bc218..adf0e618a4 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/test.hh
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/test.hh
@@ -53,7 +53,6 @@ typedef void (*testfunc_t)();	///< A unit-test function
 /// The static run() method calls all the function pointers of all instantiated
 /// objects.
 struct UnitTest {
-  static vector<UnitTest *> tests;		///< The collection of test objects
   string name;					///< Name of the test
   testfunc_t func;				///< Call-back function executing the test
 
@@ -64,7 +63,13 @@ struct UnitTest {
   UnitTest(const string &name,testfunc_t func) :
       name(name), func(func)
   {
-    tests.push_back(this);
+    tests().push_back(this);
+  }
+
+  /// \brief The collection of test objects
+  static vector<UnitTest *> & tests() {
+    static vector<UnitTest *> tests;
+    return tests;
   }
 
   static int run(set<string> &testNames);	///< Run all the instantiated tests
-- 
2.44.0

